File: src/test/kotlin/me/pcasaes/hexoids/core/domain/model/BoltTest.java
package me.pcasaes.hexoids.core.domain.model;

import me.pcasaes.hexoids.core.domain.config.Config;
import me.pcasaes.hexoids.core.domain.index.BarrierSpatialIndex;
import me.pcasaes.hexoids.core.domain.index.BarrierSpatialIndexFactory;
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndexFactory;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import pcasaes.hexoids.proto.BoltExhaustedEventDto;
import pcasaes.hexoids.proto.Event;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class BoltTest {

    @Mock
    private Clock clock;

    @Mock
    private Players players;

    private Bolts bolts;

    @BeforeEach
    public void setup() {
        MockitoAnnotations.initMocks(this);

        GameEvents.getClientEvents().registerEventDispatcher(null);
        GameEvents.getDomainEvents().registerEventDispatcher(null);

        bolts = Bolts.create();

        assertEquals(0, this.bolts.getTotalNumberOfActiveBolts());

        when(clock.getTime()).thenReturn(0L);

        Config.get().setBoltMaxDuration(10_000);
        Config.get().setBoltSpeed(0.01f);
        Config.get().setBoltCollisionRadius(0.001f);
        Config.get().setMinMove(0.000000001f);

        doAnswer(c -> Collections.emptyIterator()).when(players).iterator();
        doAnswer(c -> Collections.emptyList().spliterator()).when(players).spliterator();
        doAnswer(c -> Stream.empty()).when(players).stream();

        PlayerSpatialIndexFactory
                .factory()
                .setPlayerSpatialIndex((float x1, float y1, float x2, float y2, float distance) -> players);

        BarrierSpatialIndexFactory
                .factory()
                .setBarrierSpatialIndex(new BarrierSpatialIndex() {
                    @Override
                    public @NotNull Iterable<@NotNull Barrier> search(float x1, float y1, float x2, float y2, float distance) {
                        return Collections.emptyList();
                    }

                    @Override
                    public void update(@NotNull Iterable<Barrier> barriers) {

                    }
                });

        doAnswer(ctx -> PlayerSpatialIndexFactory.factory().get())
                .when(players)
                .getSpatialIndex();
    }

    @Test
    void testFireMoveAndExhaust() {
        List<DomainEvent> events = new ArrayList<>();
        GameEvents.getDomainEvents().registerEventDispatcher(events::add);

        EntityId one = EntityId.newId();
        Config.get().setBoltMaxDuration(1_500);
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                0f,
                0f,
                0f,
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        EntityId boltId = bolt.id;
        assertNotNull(bolt);

        assertTrue(bolt.isOwnedBy(one));

        assertEquals(1, this.bolts.getTotalNumberOfActiveBolts());


        assertEquals(1,
                bolts
                        .stream()
                        .count());

        assertTrue(bolts
                .stream()
                .anyMatch(b -> bolt == b));

        assertFalse(bolt.isExhausted());
        assertTrue(bolt.isActive());

        when(clock.getTime()).thenReturn(1_000L);

        bolts.fixedUpdate(clock.getTime());

        assertFalse(bolt.isExhausted());
        assertTrue(bolt.isActive());

        when(clock.getTime()).thenReturn(2_000L);

        events.clear();
        bolts.fixedUpdate(clock.getTime());
        assertEquals(1, events.size());
        DomainEvent domainEvent = events.get(0);
        assertNotNull(domainEvent);

        assertTrue(domainEvent.event.hasBoltExhausted());
        BoltExhaustedEventDto exhaustedEvent = domainEvent.event.getBoltExhausted();

        assertNotNull(exhaustedEvent);
        assertArrayEquals(boltId.getGuid().getGuid().toByteArray(), exhaustedEvent.getBoltId().getGuid().toByteArray());

        assertTrue(bolt.isExhausted());
        assertFalse(bolt.isActive());

        events.clear();

        assertEquals(0, bolts
                .stream()
                .count());

        assertEquals(0, this.bolts.getTotalNumberOfActiveBolts());

    }

    @Test
    void testMoveRight() {
        AtomicReference<DomainEvent> eventReference = new AtomicReference<>(null);
        GameEvents.getDomainEvents().registerEventDispatcher(eventReference::set);

        Config.get().setBoltMaxDuration(1_500);
        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                0f,
                0f,
                0f,
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);


        when(clock.getTime()).thenReturn(1_000L);

        bolts.fixedUpdate(clock.getTime());

        assertEquals(0.01f, bolt.positionVector.getX());
        assertEquals(0f, bolt.positionVector.getY());

    }

    @Test
    void testMoveRightDown() {
        AtomicReference<DomainEvent> eventReference = new AtomicReference<>(null);
        GameEvents.getDomainEvents().registerEventDispatcher(eventReference::set);

        Config.get().setBoltMaxDuration(1_500);
        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                0f,
                0f,
                (float) Math.PI / 4f,
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);


        when(clock.getTime()).thenReturn(1_000L);

        bolts.fixedUpdate(clock.getTime());

        assertEquals(0.0070710676f, bolt.positionVector.getX());
        assertEquals(0.0070710676f, bolt.positionVector.getY());
    }

    @Test
    void testMoveDown() {
        AtomicReference<DomainEvent> eventReference = new AtomicReference<>(null);
        GameEvents.getDomainEvents().registerEventDispatcher(eventReference::set);

        Config.get().setBoltMaxDuration(1_500);
        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                0f,
                0f,
                (float) Math.PI / 2f,
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);


        when(clock.getTime()).thenReturn(1_000L);

        bolts.fixedUpdate(clock.getTime());

        assertEquals(0f, bolt.positionVector.getX());
        assertEquals(0.01f, bolt.positionVector.getY());
    }

    @Test
    void testMoveLeftDown() {
        AtomicReference<DomainEvent> eventReference = new AtomicReference<>(null);
        GameEvents.getDomainEvents().registerEventDispatcher(eventReference::set);

        Config.get().setBoltMaxDuration(1_500);
        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                1f,
                0f,
                (float) (3 * Math.PI / 4f),
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);


        when(clock.getTime()).thenReturn(1_000L);

        bolts.fixedUpdate(clock.getTime());

        assertEquals(1f - 0.0070710676f, bolt.positionVector.getX());
        assertEquals(0.0070710676f, bolt.positionVector.getY());
    }

    @Test
    void testMoveLeft() {
        AtomicReference<DomainEvent> eventReference = new AtomicReference<>(null);
        GameEvents.getDomainEvents().registerEventDispatcher(eventReference::set);

        Config.get().setBoltMaxDuration(1_500);
        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                1f,
                0f,
                (float) Math.PI,
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);


        when(clock.getTime()).thenReturn(1_000L);

        bolts.fixedUpdate(clock.getTime());

        assertEquals(0.99f, bolt.positionVector.getX());
        assertEquals(0f, bolt.positionVector.getY());
    }

    @Test
    void testMoveLefUp() {
        AtomicReference<DomainEvent> eventReference = new AtomicReference<>(null);
        GameEvents.getDomainEvents().registerEventDispatcher(eventReference::set);

        Config.get().setBoltMaxDuration(1_500);
        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                1f,
                1f,
                (float) (5 * Math.PI / 4f),
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);


        when(clock.getTime()).thenReturn(1_000L);

        bolts.fixedUpdate(clock.getTime());

        assertEquals(1f - 0.0070710676f, bolt.positionVector.getX());
        assertEquals(1f - 0.0070710676f, bolt.positionVector.getY());
    }

    @Test
    void testMoveUp() {
        AtomicReference<DomainEvent> eventReference = new AtomicReference<>(null);
        GameEvents.getDomainEvents().registerEventDispatcher(eventReference::set);

        Config.get().setBoltMaxDuration(1_500);
        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                1f,
                1f,
                (float) (3 * Math.PI / 2f),
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);

        when(clock.getTime()).thenReturn(1_000L);

        bolts.fixedUpdate(clock.getTime());

        assertEquals(1f, bolt.positionVector.getX());
        assertEquals(0.99f, bolt.positionVector.getY());
    }

    @Test
    void testCollisionHit() {
        Config.get().setBoltMaxDuration(1_500);

        Player player = mock(Player.class);

        List<Player> playersList = Collections.singletonList(player);
        doAnswer(c -> playersList.iterator()).when(this.players).iterator();
        doAnswer(c -> playersList.spliterator()).when(this.players).spliterator();
        doAnswer(c -> playersList.stream()).when(this.players).stream();
        doAnswer(c -> {
            Consumer<Player> consumer = c.getArgument(0, Consumer.class);
            this.players.stream().forEach(consumer);
            return null;
        }).when(this.players).forEach(any(Consumer.class));

        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                0.5f,
                0.5f,
                0f,
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);
        EntityId boltId = bolt.id;

        doReturn(1_000L).when(clock).getTime();

        doReturn(true)
                .when(player)
                .collision(bolt.positionVector, Config.get().getBoltCollisionRadius());


        List<DomainEvent> domainEvents = new ArrayList<>();
        GameEvents.getDomainEvents().registerEventDispatcher(domainEvents::add);

        bolts.fixedUpdate(clock.getTime());

        List<Event> events = domainEvents
                .stream()
                .map(d -> d.event)
                .collect(Collectors.toList());


        BoltExhaustedEventDto boltExhaustedEventDto = events
                .stream()
                .filter(Event::hasBoltExhausted)
                .map(Event::getBoltExhausted)
                .findFirst().orElse(null);

        assertNotNull(boltExhaustedEventDto);
        assertArrayEquals(boltId.getGuid().getGuid().toByteArray(), boltExhaustedEventDto.getBoltId().getGuid().toByteArray());

    }

    @Test
    void testCollisionMiss() {
        Config.get().setBoltMaxDuration(1_500);

        Player player = mock(Player.class);

        List<Player> playersList = Collections.singletonList(player);
        doAnswer(c -> playersList.iterator()).when(this.players).iterator();
        doAnswer(c -> playersList.spliterator()).when(this.players).spliterator();
        doAnswer(c -> playersList.stream()).when(this.players).stream();

        EntityId one = EntityId.newId();
        final Bolt bolt = bolts.fired(players,
                EntityId.newId(),
                one,
                0.5f,
                0.5f,
                0f,
                Config.get().getBoltSpeed(),
                clock.getTime(),
                Config.get().getBoltMaxDuration()).orElse(null);
        assertNotNull(bolt);


        doReturn(1_000L).when(clock).getTime();
        doReturn(false)
                .when(player)
                .collision(bolt.positionVector, Config.get().getBoltCollisionRadius());

        List<DomainEvent> events = new ArrayList<>();
        GameEvents.getDomainEvents().registerEventDispatcher(events::add);

        bolts.fixedUpdate(clock.getTime());


        assertEquals(0, events
                .stream()
                .map(d -> d.event)
                .filter(Event::hasBoltExhausted)
                .count());

    }

}

File: src/test/kotlin/me/pcasaes/hexoids/core/domain/model/BoltTest.kt
package me.pcasaes.hexoids.core.domain.model

import io.mockk.every
import io.mockk.mockk
import me.pcasaes.hexoids.core.domain.config.Config
import me.pcasaes.hexoids.core.domain.index.BarrierSpatialIndex
import me.pcasaes.hexoids.core.domain.index.BarrierSpatialIndexFactory
import me.pcasaes.hexoids.core.domain.index.PlayerSpatialIndexFactory
import me.pcasaes.hexoids.core.domain.model.Bolts.Companion.create
import me.pcasaes.hexoids.core.domain.model.EntityId.Companion.newId
import me.pcasaes.hexoids.core.domain.model.GameEvents.Companion.getClientEvents
import me.pcasaes.hexoids.core.domain.model.GameEvents.Companion.getDomainEvents
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.concurrent.atomic.AtomicReference

class BoltTest {

    private lateinit var clock: Clock

    private lateinit var players: Players

    private lateinit var bolts: Bolts

    @BeforeEach
    fun setup() {
        clock = mockk(relaxed = true)
        players = mockk(relaxed = true)

        getClientEvents().registerEventDispatcher(null)
        getDomainEvents().registerEventDispatcher(null)

        bolts = create()

        Assertions.assertEquals(0, this.bolts.getTotalNumberOfActiveBolts())

        every { clock.getTime() } returns 0L

        Config.setBoltMaxDuration(10000)
        Config.setBoltSpeed(0.01F)
        Config.setBoltCollisionRadius(0.001F)
        Config.setMinMove(0.000000001F)

        every { players.iterator() } returns emptyList<Player>().iterator()
        every { players.spliterator() } returns emptyList<Player>().spliterator()

        PlayerSpatialIndexFactory
            .factory()
            .setPlayerSpatialIndex { x1: Float, y1: Float, x2: Float, y2: Float, distance: Float -> players }

        BarrierSpatialIndexFactory
            .factory()
            .setBarrierSpatialIndex(object : BarrierSpatialIndex {
                override fun search(x1: Float, y1: Float, x2: Float, y2: Float, distance: Float): Iterable<Barrier> {
                    return mutableListOf()
                }

                override fun update(barriers: Iterable<Barrier>) {
                }
            })

        every { players.getSpatialIndex() } returns PlayerSpatialIndexFactory.factory().get()
    }

    @Test
    fun testFireMoveAndExhaust() {
        val events = ArrayList<DomainEvent>()
        getDomainEvents().registerEventDispatcher { e -> events.add(e) }

        val one = newId()
        Config.setBoltMaxDuration(1500)
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            0F,
            0F,
            0F,
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        val boltId = bolt!!.id
        Assertions.assertNotNull(bolt)

        Assertions.assertTrue(bolt.isOwnedBy(one))

        Assertions.assertEquals(1, this.bolts.getTotalNumberOfActiveBolts())


        Assertions.assertEquals(
            1,
            bolts
                .count()
        )

        Assertions.assertTrue(
            bolts
                .any { b: Bolt? -> bolt == b })

        Assertions.assertFalse(bolt.isExhausted)
        Assertions.assertTrue(bolt.isActive)

        every { clock.getTime() } returns 1000L

        bolts.fixedUpdate(clock.getTime())

        Assertions.assertFalse(bolt.isExhausted)
        Assertions.assertTrue(bolt.isActive)

        every { clock.getTime() } returns 2000L

        events.clear()
        bolts.fixedUpdate(clock.getTime())
        Assertions.assertEquals(1, events.size)
        val domainEvent = events[0]
        Assertions.assertNotNull(domainEvent)
        Assertions.assertNotNull(domainEvent.event)

        Assertions.assertTrue(domainEvent.event!!.hasBoltExhausted())
        val exhaustedEvent = domainEvent.event.getBoltExhausted()

        Assertions.assertNotNull(exhaustedEvent)
        Assertions.assertArrayEquals(
            boltId.getGuid().guid.toByteArray(),
            exhaustedEvent!!.boltId.guid.toByteArray()
        )

        Assertions.assertTrue(bolt.isExhausted)
        Assertions.assertFalse(bolt.isActive)

        events.clear()

        Assertions.assertEquals(
            0, bolts
                .count()
        )

        Assertions.assertEquals(0, this.bolts.getTotalNumberOfActiveBolts())
    }

    @Test
    fun testMoveRight() {
        val eventReference = AtomicReference<DomainEvent?>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        Config.setBoltMaxDuration(1500)
        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            0F,
            0F,
            0F,
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)


        every { clock.getTime() } returns 1000L

        bolts.fixedUpdate(clock.getTime())

        Assertions.assertEquals(0.01F, bolt!!.positionVector.x)
        Assertions.assertEquals(0F, bolt.positionVector.y)
    }

    @Test
    fun testMoveRightDown() {
        val eventReference = AtomicReference<DomainEvent>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        Config.setBoltMaxDuration(1500)
        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            0F,
            0F,
            Math.PI.toFloat() / 4F,
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)


        every { clock.getTime() } returns 1000L

        bolts.fixedUpdate(clock.getTime())

        Assertions.assertEquals(0.0070710676F, bolt!!.positionVector.x)
        Assertions.assertEquals(0.0070710676F, bolt.positionVector.y)
    }

    @Test
    fun testMoveDown() {
        val eventReference = AtomicReference<DomainEvent>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        Config.setBoltMaxDuration(1500)
        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            0F,
            0F,
            Math.PI.toFloat() / 2F,
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)


        every { clock.getTime() } returns 1000L

        bolts.fixedUpdate(clock.getTime())

        Assertions.assertEquals(0F, bolt!!.positionVector.x)
        Assertions.assertEquals(0.01F, bolt.positionVector.y)
    }

    @Test
    fun testMoveLeftDown() {
        val eventReference = AtomicReference<DomainEvent>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        Config.setBoltMaxDuration(1500)
        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            1F,
            0F,
            (3 * Math.PI / 4F).toFloat(),
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)


        every { clock.getTime() } returns 1000L

        bolts.fixedUpdate(clock.getTime())

        Assertions.assertEquals(1F - 0.0070710676F, bolt!!.positionVector.x)
        Assertions.assertEquals(0.0070710676F, bolt.positionVector.y)
    }

    @Test
    fun testMoveLeft() {
        val eventReference = AtomicReference<DomainEvent>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        Config.setBoltMaxDuration(1500)
        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            1F,
            0F,
            Math.PI.toFloat(),
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)


        every { clock.getTime() } returns 1000L

        bolts.fixedUpdate(clock.getTime())

        Assertions.assertEquals(0.99F, bolt!!.positionVector.x)
        Assertions.assertEquals(0F, bolt.positionVector.y)
    }

    @Test
    fun testMoveLefUp() {
        val eventReference = AtomicReference<DomainEvent>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        Config.setBoltMaxDuration(1500)
        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            1F,
            1F,
            (5 * Math.PI / 4F).toFloat(),
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)


        every { clock.getTime() } returns 1000L

        bolts.fixedUpdate(clock.getTime())

        Assertions.assertEquals(1F - 0.0070710676F, bolt!!.positionVector.x)
        Assertions.assertEquals(1F - 0.0070710676F, bolt.positionVector.y)
    }

    @Test
    fun testMoveUp() {
        val eventReference = AtomicReference<DomainEvent>(null)
        getDomainEvents().registerEventDispatcher { newValue -> eventReference.set(newValue) }

        Config.setBoltMaxDuration(1500)
        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            1F,
            1F,
            (3 * Math.PI / 2F).toFloat(),
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)

        every { clock.getTime() } returns 1000L

        bolts.fixedUpdate(clock.getTime())

        Assertions.assertEquals(1F, bolt!!.positionVector.x)
        Assertions.assertEquals(0.99F, bolt.positionVector.y)
    }

    @Test
    fun testCollisionHit() {
        Config.setBoltMaxDuration(1500)

        val player = mockk<Player>(relaxed = true)

        val playersList = mutableListOf(player)

        every { players.iterator() } returns playersList.iterator()
        every { players.spliterator() } returns playersList.spliterator()

        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            0.5F,
            0.5F,
            0F,
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)
        val boltId = bolt!!.id

        every { clock.getTime() } returns 1000L

        every { player.collision(bolt.positionVector, Config.getBoltCollisionRadius()) } returns true


        val domainEvents = ArrayList<DomainEvent>()
        getDomainEvents().registerEventDispatcher { e -> domainEvents.add(e) }

        bolts.fixedUpdate(clock.getTime())

        val events = domainEvents
            .map { d -> d.event }


        val boltExhaustedEventDto = events
            .filter { obj -> obj!!.hasBoltExhausted() }
            .map { obj -> obj!!.getBoltExhausted() }
            .firstOrNull()

        Assertions.assertNotNull(boltExhaustedEventDto)
        Assertions.assertArrayEquals(
            boltId.getGuid().guid.toByteArray(),
            boltExhaustedEventDto!!.boltId.guid.toByteArray()
        )
    }

    @Test
    fun testCollisionMiss() {
        Config.setBoltMaxDuration(1500)

        val player = mockk<Player>(relaxed = true)

        val playersList = mutableListOf(player)

        every { players.iterator() } returns playersList.iterator()
        every { players.spliterator() } returns playersList.spliterator()


        val one = newId()
        val bolt = bolts.fired(
            players,
            newId(),
            one,
            0.5F,
            0.5F,
            0F,
            Config.getBoltSpeed(),
            clock.getTime(),
            Config.getBoltMaxDuration()
        )
        Assertions.assertNotNull(bolt)


        every { clock.getTime() } returns 1000L
        every { player.collision(bolt!!.positionVector, Config.getBoltCollisionRadius()) } returns false

        val events = ArrayList<DomainEvent>()
        getDomainEvents().registerEventDispatcher { e -> events.add(e) }

        bolts.fixedUpdate(clock.getTime())


        Assertions.assertEquals(
            0, events
                .map { d -> d.event }.count { obj -> obj!!.hasBoltExhausted() })
    }
}